flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java

```diff
@@ -205,9 +207,17 @@ public CheckpointStorageLocationReference getLocationReference() {
         return locationReference;
     }
 
+    public void registerCancelCallback(Consumer<Throwable> cancelCallback) {
+        this.cancelCallback = cancelCallback;
+    }
+
     @Override
     public void cancel(Throwable cause) {
-        targetResult.fail(cause);
+        if (cancelCallback == null) {
+            targetResult.fail(cause);
+            return;
+        }
+        cancelCallback.accept(cause);
```

Comment:
why we are not failing the `targetResult` if the `cancelCallback` is set? This seems strange.